### PR TITLE
Only allocate error when throwing, and do not subclass Error

### PR DESCRIPTION
**Pull request by scotttrinh**
There are two big changes here:

1. Only allocate an error if there is an issue instead of on every call to parse.
2. Do not "sub-class" `Error` in `ZodError`

<details>
<summary>
Currently: "object invalid 27,814 ops/sec"
</summary>

```
z.string: empty string x 48,183 ops/sec ±4.64% (82 runs sampled)
z.string: short string x 45,118 ops/sec ±5.73% (75 runs sampled)
z.string: long string x 50,951 ops/sec ±3.25% (81 runs sampled)
z.string: invalid: null x 36,886 ops/sec ±3.92% (84 runs sampled)
z.string: manual parser: long x 700,950,412 ops/sec ±3.64% (84 runs sampled)
z.object: empty: valid x 25,550 ops/sec ±3.01% (79 runs sampled)
z.object: empty: valid: extra keys x 25,125 ops/sec ±2.78% (86 runs sampled)
z.object: empty: invalid: null x 36,037 ops/sec ±3.59% (84 runs sampled)
z.object: short: valid x 13,711 ops/sec ±4.84% (80 runs sampled)
z.object: short: valid: extra keys x 13,239 ops/sec ±4.70% (80 runs sampled)
z.object: short: invalid: null x 34,528 ops/sec ±7.57% (81 runs sampled)
z.object: long: valid x 7,490 ops/sec ±3.97% (63 runs sampled)
z.object: long: valid: extra keys x 7,673 ops/sec ±4.08% (84 runs sampled)
z.object: long: invalid: null x 27,814 ops/sec ±8.56% (64 runs sampled)
```

</details>

<details>
<summary>
Only allocate on failure: "object invalid 27,761 ops/sec"
</summary>

```
z.string: empty string x 560,109 ops/sec ±6.24% (80 runs sampled)
z.string: short string x 573,299 ops/sec ±5.57% (80 runs sampled)
z.string: long string x 608,116 ops/sec ±4.59% (88 runs sampled)
z.string: invalid: null x 34,280 ops/sec ±4.15% (62 runs sampled)
z.string: manual parser: long x 705,877,940 ops/sec ±4.06% (86 runs sampled)
z.object: empty: valid x 45,039 ops/sec ±4.72% (66 runs sampled)
z.object: empty: valid: extra keys x 46,582 ops/sec ±5.01% (77 runs sampled)
z.object: empty: invalid: null x 26,996 ops/sec ±16.14% (63 runs sampled)
z.object: short: valid x 31,774 ops/sec ±5.54% (73 runs sampled)
z.object: short: valid: extra keys x 34,718 ops/sec ±5.30% (76 runs sampled)
z.object: short: invalid: null x 32,920 ops/sec ±8.31% (75 runs sampled)
z.object: long: valid x 17,128 ops/sec ±16.67% (64 runs sampled)
z.object: long: valid: extra keys x 23,249 ops/sec ±5.94% (76 runs sampled)
z.object: long: invalid: null x 27,761 ops/sec ±8.13% (69 runs sampled)
```

</details>

<details>
<summary>
Do not subclass Error: "object invalid 86,946 ops/sec"
</summary>

```
z.string: empty string x 540,451 ops/sec ±3.46% (86 runs sampled)
z.string: short string x 348,678 ops/sec ±17.25% (59 runs sampled)
z.string: long string x 436,669 ops/sec ±15.48% (82 runs sampled)
z.string: invalid: null x 111,825 ops/sec ±3.28% (86 runs sampled)
z.string: manual parser: long x 720,663,528 ops/sec ±3.03% (86 runs sampled)
z.object: empty: valid x 388,305 ops/sec ±4.97% (88 runs sampled)
z.object: empty: valid: extra keys x 306,263 ops/sec ±13.89% (75 runs sampled)
z.object: empty: invalid: null x 98,812 ops/sec ±14.33% (77 runs sampled)
z.object: short: valid x 153,457 ops/sec ±9.91% (79 runs sampled)
z.object: short: valid: extra keys x 149,708 ops/sec ±5.59% (83 runs sampled)
z.object: short: invalid: null x 111,474 ops/sec ±5.10% (81 runs sampled)
z.object: long: valid x 77,332 ops/sec ±3.48% (86 runs sampled)
z.object: long: valid: extra keys x 73,265 ops/sec ±4.34% (86 runs sampled)
z.object: long: invalid: null x 86,946 ops/sec ±12.95% (74 runs sampled)
```

</details>

<details>
<summary>
With both changes: "object invalid 121,532 ops/sec"
</summary>

```
z.string: empty string x 580,162 ops/sec ±4.84% (83 runs sampled)
z.string: short string x 576,156 ops/sec ±5.78% (82 runs sampled)
z.string: long string x 515,700 ops/sec ±9.86% (76 runs sampled)
z.string: invalid: null x 95,510 ops/sec ±6.18% (77 runs sampled)
z.string: manual parser: long x 689,104,213 ops/sec ±3.86% (82 runs sampled)
z.object: empty: valid x 427,090 ops/sec ±6.95% (88 runs sampled)
z.object: empty: valid: extra keys x 374,931 ops/sec ±4.91% (79 runs sampled)
z.object: empty: invalid: null x 114,452 ops/sec ±4.92% (81 runs sampled)
z.object: short: valid x 178,976 ops/sec ±3.61% (82 runs sampled)
z.object: short: valid: extra keys x 169,163 ops/sec ±3.80% (85 runs sampled)
z.object: short: invalid: null x 112,253 ops/sec ±3.63% (87 runs sampled)
z.object: long: valid x 78,291 ops/sec ±4.50% (88 runs sampled)
z.object: long: valid: extra keys x 81,841 ops/sec ±2.74% (90 runs sampled)
z.object: long: invalid: null x 121,532 ops/sec ±3.58% (90 runs sampled)
```

</details>

### CODE REVIEW

File: deno/lib/types.ts
Comment (lilred) [inline deno/lib/types.ts:90]: @colinhacks I wish this function didn't signal failure by throwing since there's nothing in principle exceptional about validation failing... There could be performance impact in either direction.

I'm not sure if you agree that this is worth investigating, if so this could be a good area for future work (after this PR).

File: deno/lib/types.ts
Comment (lilred) [inline deno/lib/types.ts:1666]:
```suggestion
      (): ZodIssue[] => []
```

**Review by lilred — CHANGES_REQUESTED** [review]
That's it for me. Thank you so much for doing this!

File: deno/lib/types.ts
Comment (scotttrinh) [reply deno/lib/types.ts]: @lilred ⚡ [`51d9e9c` (#459)](https://github.com/colinhacks/zod/pull/459/commits/51d9e9c2d823ef2cfdda86756cd7d85cdf186068)

**Comment by scotttrinh** [comment]
Note that this also includes a bunch of benchmarking tooling (from #431 ), so if we want to kick that out I can remove it once we're ready to merge.

**Comment by lilred** [comment]
I'm very eager to add benchmarking and it looks like you did a very good job to get the ball rolling so I wouldn't remove it.

**Review by lilred — APPROVED** [review]

**Comment by scotttrinh** [comment]
@colinhacks Updated with latest changes. Any thoughts/criticisms of this approach? I know a lot of us are anxious to put this in our production systems but can't take the performance hit.

**Comment by rossholdway** [comment]
This looks great 👍  Zod is great, I'm a sponsor - but I moved away due to performance when compared with similar libraries. It'd be great to see this merged!

**Comment by colinhacks** [comment]
Sorry, very late in getting to this. Thanks for the bump @rossholdway. 

There's an argument for subclassing Error. It lets us maintain stack traces and avoid breaking any `instanceof Error` logic implemented by Zod users. @lilred you expressed this same thought [here](https://github.com/colinhacks/zod/pull/292#discussion_r567142902). Have you changed your stance on this? I could be convinced either way I suppose.

Another thought: changing the type of `ZodInvalidUnionIssue.unionIssues` from `ZodError` to `ZodIssue[][]` might be unnecessarily disruptive. If we changed it from `ZodIssue[][]` to `{ issues: ZodIssue[] }[]` then it would break fewer codebases. Though union schemas are rare enough that I think keeping it as `ZodError[]` to avoid a breaking change is probably worth the performance cost.

**Comment by colinhacks** [comment]
@scotttrinh This is a fantastic PR, thanks for all your work here! I should have said that much earlier. Amazing stuff. Exactly how I would have done it. 

I don't have a problem with the changes to `ZodInvalidUnionIssue` - your format makes way more sense. I also think getting rid of the `Error` subclassing is fine. The stack trace captured [a bunch of unnecessary information](https://github.com/colinhacks/zod/issues/101) anyway. I'm likely to merge #492 after folding in some of your improvements here. If you have any thoughts on that PR I'd love to hear them!

**Comment by lilred** [comment]
>@lilred you expressed this same thought here. Have you changed your stance on this? I could be convinced either way I suppose.

Hard requirement: `safeParse` must not instantiate an `Error` ever
Soft requirement: `parse` should instantiate an `Error` upon failure

I think it you're using a `try/catch` flow you probably expect an `Error` to come out. That being said this is not an opinion I hold strongly.

>Though union schemas are rare enough that I think keeping it as ZodError[] to avoid a breaking change is probably worth the performance cost.

I use union schemas all the time in API validation. Perf certainly matters. I like `{ issues: ZodIssue[] }[]` because it should break marginally less client code at ~no performance cost.

That being said I want to give deference to @scotttrinh on this, he's been on this for a while and better a flawed PR merged than no PR merged. @scotttrinh, if you have the patience and inclination to implement these comments I would be super grateful. Otherwise I'll defer to you and @colinhacks.

**Comment by scotttrinh** [comment]
> That being said I want to give deference to @scotttrinh on this, he's been on this for a while and better a flawed PR merged than no PR merged. @scotttrinh, if you have the patience and inclination to implement these comments I would be super grateful. Otherwise I'll defer to you and @colinhacks.

I'm more than happy to work on addressing these comments, shouldn't be that big of a lift. I wonder if we should work on this in the context of #492 or keep them separate for now? I'd rather spend more time to get a 100x increase then spend time on getting to 10x sooner.

@colinhacks I'll follow your lead here and help get this and any other perf PRs in a mergeable state, just let me know how best to support you on it. I suspect that there will be a lot of conflicts if we're both mucking around in `types.ts` making slightly incompatible changes.

**Comment by scotttrinh** [comment]
Since #492 landed, I think we can close this. Huge shout out to @reinismu for the initial work on this specific optimization, and @milankinen for the amazing work on the (more fundamental) change that actually landed.

**Comment by colinhacks** [comment]
Thanks @scotttrinh and @reinismu! Incredible work.

**Pull request closed without merging**
